builtin/logical/aws/path_roles.go

```diff
@@ -261,6 +268,16 @@ func (b *backend) pathRolesWrite(ctx context.Context, req *logical.Request, d *f
 		roleEntry.ProhibitFlexibleCredPath = false
 	}
 
+	if _, ok := d.GetOk("user_path"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with user_path"), nil
+		}
+		if !strutil.StrListContains(roleEntry.CredentialTypes, iamUserCred) {
+			return logical.ErrorResponse(fmt.Sprintf("user_path parameter only valid for %s credential type", iamUserCred)), nil
+		}
+	}
+	roleEntry.UserPath = d.Get("user_path").(string)
```

Comment:
This is the wrong pattern here. Look at how all the other calls to `GetOk` are handled. What this code will do is, if a user wants to update an existing role and does NOT specify the `user_path` option, this will always reset it to the default of "/" which isn't correct.

The normal pattern is to do something like `if user_path_raw := d.GetOk("user_path"); ok {` and then inside that if block, handle the value for `user_path_raw`. Then, if needed, there would be another check to see if the operation were a `CreateOperation` and, if so, supplying the default value. However, this endpoint doesn't implement a listener for a `CreateOperation` and I think it's simplest if it continues to not and instead has `secret_access_keys.go` handle this (which it probably needs to).

Reply:
The pattern `if user_path_raw, ok := d.GetOk("user_path"); ok {` is how I did at first and then I got confused by legacy role. Because if legacy role is not empty it will ignore `user_path` default value `/`  and will set empty string which seemed a bit wrong. So i thought to set always to default `/` if not provided.  And regarding your concern about reseting `user_path` if not provided - isn't that expected behaviour? Or does aws backend don't follow kv conventions?

Reply:
My suggestion would be to disallow setting `user_path` if it's a legacy role. (Maybe a whitelisted set of parameters if it is a legacy role, to make this easier in the future?) If people want the new features, they should upgrade to a non-legacy role :)

And see my above comments about resetting `user_path` if not provided.

Reply:
But still the problem persists if I we dont provide `user_path` for the very first time when we create this path. If we dont use default value `/` it will write empty string whereas IMO it should use write `/`, no?

Reply:
It depends on how you address https://github.com/hashicorp/vault/pull/6380#discussion_r264498698 -- if you gracefully handle the `user_path` being an empty string there, then no -- you can leave an empty string here and it'll be handled correctly. If you implement role upgrading logic, then yes, it'll need to be set to `/`.